### ISSUE TITLE
Emscripten fetch API won't invoke callback if the calling thread is in waiting.

### ISSUE THREAD

**Issue by keminming** [emscripten-core/emscripten#13752]
We ran into an issue with Emscripten fetch API in ASYNC mode. 
After a thread issued a Emscripten fetch call, we put it in waiting status, meaning we call std::conditional_variable::wait() on the calling thread. In this case, the fetch callbacks never get invoked.

The structure of our code is something like:

- function definition:

```
void invokeNetworkingCall(callback){ 
 _attr.onsuccess = callback;
 emscripten_fetch(_attr, ...);
}
```
/////////////////////////////////////////////

- In the calling thread:
```

std::conditional_variable c;
invokeNetworkingCall([&c]{
  // some operations
   c.notify_all();
});
c.wait();
```

I've read the documentation about the "Waitable" mode, but not sure if that's related to this issue. But anyway, I wonder what's the reason this happens, even I've pass the -pthread + -s PROXY_TO_PTHREAD to the compiler/linker, which enable pthread and ran app on it.  Shouldn't the callback of "Fetch API" triggered from a different thread than the calling one?

Thanks!

**Comment by sbc100** [comment]
Indeed, I think there is no way for the fetch response to arrive if you are blocking the thread like that.

The same goes for waiting on any kind of web event (e..g setTimeout, RAF, etc).   If you block the main thread none of those events can be delivered to you.

In order to receive the even you need to run the event loop somehow.  Normally this is don't by simply returns out of your callstack, but you can also use ASYNCIFY if you really can't do that (https://emscripten.org/docs/porting/asyncify.html).

Alternatively you could run your fetch requests on a background pthread.

**Comment by keminming** [comment]
Ok, we are not able to use ASYNCIFY, so we want to run fetch on a background pthread. So clarify, can I do something like:
In main:
```
std::conditional_variable c;
std::thread t{ [&c]{ invokeNetworkingCall([&c]{
  // some operations
   c.notify_all();
});
c.wait();
```

**Comment by sbc100** [comment]
I think so yes.  I'm not too familiar with the C++ way of doing things but you should be able to start a thread and have that thread do all the fetches.. as long as the thread you start is not itself blocked on anything.   The main thread can then block on he fetch thread.   

The ability to then join the fetch thread is something I recently add/improved in #13596.   This change has not yet made it into a release so you would need to `./emsdk install tot` to get access to it.

Of course blocking the main thread at all on the web is not really advisable (and in fact emscripten has to fake it by busy looping).. is it actually the main thread you want to block or are you already on a background thread?

**Comment by keminming** [comment]
Yeh, I think our overall product is already on background thread. For my small component, I created a simply test with -s PROXY_TO_PTHREAD flags, then I will call wait/block on the main function(what I believe it's already on a secondary thread?)

So curious which API in emscripten could cause main to block? 
And by taking #13596, does it mean we can do something like:
in main:
```
std::thread t{ []{ invokeNetworkingCall([]{
  // some operations
});
t.join()
```

But what if we want the main function to continue only after the fetch callback get invoked rather than the fetch thread exit?

Thanks.

**Comment by sbc100** [comment]
> Yeh, I think our overall product is already on background thread. For my small component, I created a simply test with -s PROXY_TO_PTHREAD flags, then I will call wait/block on the main function(what I believe it's already on a secondary thread?)
> 
> So curious which API in emscripten could cause main to block?

When I say "block the main thread" I mean block the main browser thread.  If you are already using another thread (.e.g if you are using `PROXY_TO_PTHREAD` you don't need to worry about that.   If you are running code directly on the main browser thread any kind of `.join` on a thread or or `.wait` on a mutex is potentially dangerous and no advisable.

`.join` or `.wait` on a background thread on prevents JS events arriving on that thread.   But `.join` or `.wait` on the main browser thread can cause the entire page to hang and eventually be killed.

> And by taking #13596, does it mean we can do something like:
> in main:
> 
> ```
> std::thread t{ []{ invokeNetworkingCall([]{
>   // some operations
> });
> t.join()
> ```
> 
> But what if we want the main function to continue only after the fetch callback get invoked rather than the fetch thread exit?

#13596 means that the fetch thread itself will not be join-able (will not really exit per-say) until all the fetches are done.  It is kept along until all its async work is done then it becomes joinable.. at least that is the idea.

> 
> Thanks.

**Comment by keminming** [comment]
Thanks. The information is very useful and I feel much more clear now.

**Comment by keminming** [comment]
Yesterday I tried one more thing, instead of calling fetch from a new thread, I dispatch the call to an executor(thread pool). The reason is we tried to avoid the overhead frequently construct/destruct new thread. This time I found the callback never get invoked. 

Like to know if this is an expected behavior?

**Comment by sbc100** [comment]
As long as the thread that makes the call does not block the callback should arrive on that same thread.    

Threads should not exit until the callback completes (even though the can/must return from their entry points to recieve the callback).   Once the callback is executed (always on the same thread that made the call) then the thread can exit when callback returns (assuming not other outstanding work).

**Comment by keminming** [comment]
Ok, I think with thread pool, the thread would inevitable be blocked exiting the routine calling fetch, in order to wait for the next task. So I guess the answer for thread pool is "NO".

Think I will try with the SYNC mode, so that it won't have the chance to block event I guess.

**Comment by sbc100** [comment]
Whatever thread kicks of the fetch must be structured in such as way that it can receive JS events.. otherwise the callback will never arrive.  The only way to receive JS events is to return the event loop (at least occasionally). 

Alternatively if you are on a worker there maybe some way to do a blocking (synchronous) fetch that does not involve a callback at all and therefore does't need event processing to run.
